### PR TITLE
[VSX] Display plugins counts

### PR DESCRIPTION
**Pull request by alvsan09**
#### What it does

Adds a plugin count badge to the title of each instance of
`VSXExtensionsWidget` except for `SEARCH_RESULT` as the count may be
partial.

The solution extends the functionality under `view-container` by adding a `BadgeWidget` interface that can be
implemented by widgets wrapped in a `ViewContainerPart`, this is the reference implementation for it.
This solution does not show the badge for view parts that contain a toolbar populated with more than one item.

#### How to test

- Add an `extensions.json` file with the following content to your workspace `.theia` folder:
```ts
{
    "recommendations": [
        "ms-azuretools.vscode-docker",
        "exiasr.hadolint",
        "timonwong.shellcheck",
        "coenraads.bracket-pair-colorizer-2",
        "scalameta.metals"
    ]
}
```
- Start `theia` and open the `EXTENSIONS` view
- Install few extensions
- Verify the counter values for each widget i.e. `INSTALLED`, `BUILT-IN`, `RECOMMENDED`
- As per the listed extensions above the expected count for `RECOMMENDED` is **5**
- Make sure other view parts don't display a badge e.g. `WATCH`  under the `DEBUG` view.

vsx_plugin_count gif

#### Review checklist

- [x] As an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- As a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

### CODE REVIEW

**Review by msujew — COMMENTED** [review]
Looks good from a surface viewpoint. Haven't looked into the code yet. I noticed some minor layouting issues, where the badge clips into the focus outline. I believe we need a larger margin there:

grafik

VSCode:

grafik

**Comment by alvsan09** [comment]
> Looks good from a surface viewpoint. Haven't looked into the code yet. I noticed some minor layouting issues, where the badge clips into the focus outline. I believe we need a larger margin there:
> 
> grafik
> 
> VSCode:
> 
> grafik

I also received some comments off-line from @vince-fugnitto , I am in the process to change the code, i.e.
* Give the responsibility to render the badge to the `ViewContainerPart` rather than from the `VSXExtensionsWidget`
* Make the `BadgeWidget` interface more specific to contain a `badge` of type `number | undefined` rather that `node | undefined`
* Make sure that if a part contains a `toolbar` the widget is not displayed.

I will add a right margin to it as you have pointed out as well, 
Additional comments are very welcome !
Thanks @msujew !

File: packages/core/src/browser/style/notification.css [hunk 1]
```diff
@@ -34,6 +34,10 @@
     text-align: center;
 }
 
+.notification-count-container-margin {
```
Comment (vince-fugnitto) [inline packages/core/src/browser/style/notification.css:37]: We should drop `-margin` from the classname.

Comment (vince-fugnitto) [inline packages/core/src/browser/style/notification.css:37]: I believe this file is strictly for notifications styling, shouldn't we put our styling under `view-container.css`?
The use of `notifications` for badge styling in `view-container` seems a little odd.

File: packages/vsx-registry/src/browser/vsx-extensions-widget.ts [hunk 2]
```diff
@@ -57,27 +59,60 @@ export class VSXExtensionsWidget extends SourceTreeWidget {
...
             default:
-                return '';
+                label = '';
+        }
+        const title = `${label}`;
+        this.title.label = title;
+        this.title.caption = title;
+
+        this.badge = await this.resolveCount();
+    }
```
Comment (vince-fugnitto) [inline packages/vsx-registry/src/browser/vsx-extensions-widget.ts:108]: I believe this method is doing too much work and merges two concepts together (which is confusing), the `title` of the view-part and handling the `badge`.

File: packages/vsx-registry/src/browser/vsx-extensions-widget.ts [hunk 3]
```diff
@@ -57,27 +59,60 @@ export class VSXExtensionsWidget extends SourceTreeWidget {
...
+
+    protected async resolveCount(): Promise<number | undefined> {
+        if (this.options.id !== VSXExtensionsSourceOptions.SEARCH_RESULT) {
+            const elements = await this.source?.getElements() || [];
+            return [...elements].length;
         }
+        return undefined;
```
Comment (vince-fugnitto) [inline packages/vsx-registry/src/browser/vsx-extensions-widget.ts:115]: The changes do not properly handle a `0` count:

image

In this case I would still expect `0` to be displayed for recommended and installed.

Comment (alvsan09) [reply packages/vsx-registry/src/browser/vsx-extensions-widget.ts:115]: I have submitted a new commit to address these comments !!
Thanks @vince-fugnitto !!

File: packages/vsx-registry/src/browser/vsx-extensions-widget.ts [hunk 4]
```diff
@@ -52,32 +54,65 @@ export class VSXExtensionsWidget extends SourceTreeWidget {
     protected readonly extensionsSource: VSXExtensionsSource;
 
     @postConstruct()
-    protected override init(): void {
+    protected override async init(): Promise<void> {
```
Comment (paul-marechal) [inline packages/vsx-registry/src/browser/vsx-extensions-widget.ts:57]: Don't make the post-construct method async: Inversify 5 doesn't wait for it.

Please use the `Promise.then` API instead.

File: packages/vsx-registry/src/browser/vsx-extensions-widget.ts [hunk 5]
```diff
@@ -57,14 +59,31 @@ export class VSXExtensionsWidget extends SourceTreeWidget {
...
         this.title.caption = title;
 
-        this.toDispose.push(this.extensionsSource);
-        this.source = this.extensionsSource;
+        this.toDispose.push(this.source.onDidChange(async () => {
+            this.badge = await this.resolveCount();
+        }));
+    }
+
+    private _badgeCount: number | undefined = undefined;
```
Comment (vince-fugnitto) [inline packages/vsx-registry/src/browser/vsx-extensions-widget.ts:75]: minor: we should rename to `_badge` due to the getter and setter, or just use `badgeCount` (dropping the unnecessary `_`). I'm in favor of using `_badge` however.

- https://github.com/eclipse-theia/theia/wiki/Coding-Guidelines#no_underscore_private

**Review by vince-fugnitto — COMMENTED** [review]
I confirm that the changes work as expected:

- [x] a badge of `0` is supported and displayed
- [x] the 'builtin', 'recommended' and 'installed' tree-view-parts have badges
- [x] the badges correctly display the extension count for each part
- [x] the badges are correctly updated when the count is updated

File: packages/vsx-registry/src/browser/vsx-extensions-widget.ts (hunk 5 above)
Comment (alvsan09) [reply packages/vsx-registry/src/browser/vsx-extensions-widget.ts]: I agree with you, `_badge` is more appropriate, 
It has been updated in the latest commit, 
I have also squashed the commits and re-based it to later master.

Thanks @vince-fugnitto !

**Review by vince-fugnitto — APPROVED** [review]
LGTM :+1:

**Comment by alvsan09** [comment]
@msujew, @vince-fugnitto, Thanks !! :+1:

**Pull request merged**